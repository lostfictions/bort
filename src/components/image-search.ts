import axios from "axios";
import cheerio from "cheerio";

import { randomInt } from "../util";

/**
 * Perform an image search, optionally filter recently-seen images, and select a
 * random image from the top N remaining results.
 */
export async function imageSearch({
  term,
  selectFromTop = 5,
  recents,
  animated = false,
}: {
  term: string;
  selectFromTop?: number;
  recents?: { [url: string]: unknown };
  animated?: boolean;
}): Promise<string | false> {
  let res = await requestAndParse({ term, animated, exact: true });
  if (res.length === 0) {
    // if no results, try an inexact search
    res = await requestAndParse({ term, animated });
  }

  let cursor = 0;
  let sliced: string[] = [];

  while (sliced.length === 0 && cursor < res.length) {
    sliced = res.slice(cursor, cursor + selectFromTop);

    if (recents) {
      for (let i = sliced.length - 1; i >= 0; i--) {
        // TODO [-level] replace `in` operator
        // eslint-disable-next-line no-restricted-syntax
        if (sliced[i] in recents) {
          sliced.splice(i, 1);
        }
      }
    }

    cursor += selectFromTop;
  }

  while (sliced.length > 0) {
    let [result] = sliced.splice(randomInt(sliced.length), 1);

    // rewrite .gifv => .gif
    if (result.endsWith(".gifv")) {
      result = result.slice(0, -1);
    }

    // test that it actually exists
    try {
      await axios.head(result);
      return result;
    } catch {
      // don't care, just try another result
    }
  }

  return false;
}

export interface ImageSearchOptions {
  term: string;
  animated?: boolean;
  exact?: boolean;
  transparent?: boolean;
}

export async function requestAndParse(options: ImageSearchOptions) {
  const res = await request(options);
  return parse(res.data);
}

export function request({
  term,
  animated,
  exact,
  transparent,
}: ImageSearchOptions) {
  const tbs = [transparent && "ic:trans", animated && "itp:animated"]
    .filter(Boolean)
    .join(",");

  return axios.get<string>("https://www.google.com/search", {
    params: {
      q: term,
      tbm: "isch",
      nfpr: exact ? 1 : 0,
      tbs: tbs || undefined,
    },
    timeout: 5000,
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 " +
        "(KHTML, like Gecko) Chrome/79.0.3945.79 Safari/537.36",
    },
  });
}

export function parse(html: string) {
  const $ = cheerio.load(html);

  const strategies = [allJsonpStrategy];

  for (const s of strategies) {
    const res = s($);
    if (res && res.length > 0) return res;
  }

  return [];
}

const PREFIX = "AF_initDataCallback";

export function allJsonpStrategy($: cheerio.Root): string[] | false {
  const scripts = $("script")
    .toArray()
    .map((el) => $(el).text())
    .filter((t) => t.startsWith(PREFIX));

  if (scripts.length > 0) {
    return scripts
      .flatMap((script) => [
        ...script.matchAll(/"(https?:\/\/[^"]+\.(?:jpe?g|gifv?|png))"/g),
      ])
      .map((res) =>
        // google image search results sometimes contain code points encoded as
        // eg. \u003d, so replace them with their actual values
        res[1].replace(/\\u([a-fA-F0-9]{4})/gi, (_, g) =>
          String.fromCodePoint(parseInt(g, 16))
        )
      );
  }

  return false;
}
