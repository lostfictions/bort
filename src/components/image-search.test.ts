import { readFileSync, readdirSync } from "fs";
import { join } from "path";

import { imageSearch, parse } from "./image-search";

const fixturePath = join(__dirname, `../../fixtures/image-search`);
const fixtureFiles = readdirSync(fixturePath);

const fixtures = fixtureFiles.map(f => [
  readFileSync(join(fixturePath, f), "utf8"),
  f
]);

// const mockResponses = {
//   "1": loadFixture(1),
//   "2": loadFixture(2),
//   "3": loadFixture(3)
// };

// jest.mock("axios", () =>
//   jest.fn(async (url: string) => {
//     if (url.endsWith("1")) return { data: mockResponses["1"] };
//     if (url.endsWith("2")) return { data: mockResponses["2"] };
//     if (url.endsWith("3")) return { data: mockResponses["3"] };
//     throw new Error(`Unexpected url calling mocked axios: ${url}`);
//   })
// );

describe("image search", () => {
  describe("parse", () => {
    it("matches snapshots from fixture data", () => {
      fixtures.forEach(([fixture, fixtureName]) => {
        expect(parse(fixture)).toMatchSnapshot(fixtureName);
      });
    });
  });
});
