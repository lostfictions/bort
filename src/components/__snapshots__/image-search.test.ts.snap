// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image search parse matches snapshots from fixture data: 01-puppies.html 1`] = `
Array [
  "https://d17fnq9dkz9hgj.cloudfront.net/uploads/2020/04/shelter-dog-cropped-1-632x329.jpg",
  "https://dogtime.com/assets/uploads/2018/10/puppies-cover-1280x720.jpg",
  "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/dog-puppy-on-garden-royalty-free-image-1586966191.jpg",
  "https://i.cbc.ca/1.5507480.1585002199!/fileImage/httpImage/puppy-facts.jpg",
  "https://cdn2-www.dogtime.com/assets/uploads/2016/09/rottweiler-puppy-1-e1553630142158.jpg",
  "http://www.iams.com/images/default-source/article-image/article_puppy-basics-selecting-the-right-food_header.jpg",
  "http://cdn.akc.org/content/article-body-image/siberian_husky_cute_puppies.jpg",
  "https://www.thesprucepets.com/thmb/GBpggbhEYcY8glHE6smP0PIZY4Q\\\\u003d/2121x1414/filters:fill(auto,1)/golden-retriever-puppy-in-grass-923135452-5c19243546e0fb000190737c.jpg",
  "https://cdn.theatlantic.com/assets/media/img/photo/2020/03/photos-national-puppy-day-2020/p01_1490227862-1/original.jpg",
  "https://thehappypuppysite.com/wp-content/uploads/2018/04/when-can-puppies-go-outside-header.jpg",
  "https://hips.hearstapps.com/countryliving.cdnds.net/17/47/1511194376-cavachon-puppy-christmas.jpg",
  "https://s3.amazonaws.com/cdn-origin-etr.akc.org/wp-content/uploads/2017/11/25233546/Petit-Basset-Griffon-Vendeen-puppy-laying-indoors.jpg",
  "https://www.thesprucepets.com/thmb/91o3ucLXmSkvVPfWYBQOohQk25A\\\\u003d/4707x3530/smart/filters:no_upscale()/puppies-sharing-an-ice-cream-cone-991329222-5b9ae36a46e0fb0025d890b1.jpg",
  "https://media.mnn.com/assets/images/2019/11/narwhal_puppy_tail_forehead.jpg.653x0_q80_crop-smart.jpg",
  "https://www.vicnews.com/wp-content/uploads/2019/12/19954317_web1_puppies-3.jpg",
  "https://www.readersdigest.ca/wp-content/uploads/2013/03/6-facts-to-know-before-owning-a-puppy.jpg",
  "https://dogtime.com/assets/uploads/2011/03/puppy-development-1280x720.jpg",
  "https://9ed48207422fa7fc5013-a6297eb5ec0f30e883355c8680f3b2d6.ssl.cf2.rackcdn.com/35585131450_6-20190322154216-20190322154226-1000x1000.jpg",
  "https://i.ytimg.com/vi/rziIg5V1RdA/maxresdefault.jpg",
  "https://www.saanichnews.com/wp-content/uploads/2019/06/17219549_web1_news-abandoned-puppies-quesnel.jpg",
  "https://media.mnn.com/assets/images/2018/05/puppies_dog_bed.jpg.1440x960_q100_crop-scale_upscale.jpg",
  "https://www.vettedpetcare.com/vetted-blog/wp-content/uploads/2019/09/cute-corgi-puppy.jpg",
  "https://i.ytimg.com/vi/dzG1TL3CjEs/maxresdefault.jpg",
  "https://beta.ctvnews.ca/local/british-columbia/2020/4/13/1_4894930/_jcr_content/root/responsivegrid/image.coreimg.png/1586834472634/1-4894935.png",
  "https://blog.malwarebytes.com/wp-content/uploads/2017/11/shutterstock_688527661-900x506.jpg",
  "https://upload.wikimedia.org/wikipedia/commons/6/6e/Golde33443.jpg",
  "https://ovc.uoguelph.ca/sites/default/files/users/k.mantel/images/iStock-467825446%20cute%20labador%20puppy%202.jpg",
  "http://cdn.akc.org/content/article-body-image/rottie_cute_puppies.jpg",
  "https://www.caninefitness.com/images/rottweiler-puppies.jpg",
  "https://thumbs-prod.si-cdn.com/6J8x6o4CS1HIZ8PDcxmy4jXVNtE\\\\u003d/fit-in/1600x0/filters:focal(1501x569:1502x570)/https://public-media.si-cdn.com/filer/1e/1e/1e1e3aea-e930-45d2-9bc4-6dd50cccf66d/istock-511313058.jpg",
  "https://www.allivet.com/blog/wp-content/uploads/2016/03/puppies-688425_1920-min.jpg",
  "https://www.akcpetinsurance.com/res/akc/blog/2018/new-puppy-essentials/header_puppy_checklist.jpg",
  "https://www.hillspet.com/content/dam/cp-sites/hills/hills-pet/en_us/exported/dog-care/new-pet-parent/images/golden-retriever-puppy-running-outside.jpg",
  "https://www.vernonmorningstar.com/wp-content/uploads/2018/12/14627357_web1_181205-VMS-bernese-puppy.jpg",
  "https://www.thesprucepets.com/thmb/_cQioAhFn0ly0v_aDgi1quRiVac\\\\u003d/3168x2109/filters:fill(auto,1)/puppy-samoyed-boy-990077480-5c89719646e0fb00012c67e8.jpg",
  "https://i.ytimg.com/vi/a6KGPBflhiM/hqdefault.jpg",
  "https://i.cbc.ca/1.5117915.1556720294!/fileImage/httpImage/image.jpg_gen/derivatives/16x9_780/abbotsford-police-stolen-puppies-twitter.jpg",
  "https://www.readersdigest.ca/wp-content/uploads/2017/08/train-your-puppy-how-to-sit.jpg",
  "https://www.cp24.com/polopoly_fs/1.4937754.1589389981!/httpImage/image.jpg_gen/derivatives/landscape_620/image.jpg",
  "https://www.rd.com/wp-content/uploads/2020/02/GettyImages-1050600824-1024x683.jpg",
  "https://exclusive.multibriefs.com/images/exclusive/0513stbernardpuppy.jpg",
  "https://ichef.bbci.co.uk/news/1024/branded_news/D35E/production/_106801145_puppies.jpg",
  "https://www.thelabradorsite.com/wp-content/uploads/2017/05/When-can-puppies-go-outside-LS-long.jpg",
  "https://compote.slate.com/images/24c26e48-a422-4843-90c3-927c3268f51e.jpg",
  "https://images-na.ssl-images-amazon.com/images/I/717I6AVyItL.jpg",
  "https://www.citynews1130.com/wp-content/blogs.dir/sites/9/2020/04/09/Port-alberni-puppies-scaled.jpg",
  "https://www.coastmountainnews.com/wp-content/uploads/2018/09/13555856_web1_180919-VNE-VictoriaHumanePuppies.jpg",
  "https://www.gannett-cdn.com/presto/2019/12/18/USAT/5c51582a-64cb-4079-8584-b7bdf9dae701-GettyImages-1130714351.jpg",
  "https://thestacker.com/sites/default/files/styles/properly_sized_image/public/2020-03/English%20Lab%20Puppy%20%281%29_0.png",
  "https://cdn.cnn.com/cnnnext/dam/assets/191204034536-puppies-rescued-2.jpg",
  "https://urbananimalveterinary.com/wp-content/uploads/2019/11/puppy.jpg",
  "https://www.thechronicleherald.ca/media/photologue/photos/cache/pei-humane-society-begins-process-of-giving-husky-puppies-forever-homes-2_large.jpg",
  "https://i.dailymail.co.uk/1s/2019/12/18/14/22403936-7805893-image-a-1_1576680786169.jpg",
  "https://uwaterloo.ca/housing/sites/ca.housing/files/resize/uploads/images/puppies-340x220.jpg",
  "https://s3.amazonaws.com/cdn-origin-etr.akc.org/wp-content/uploads/2017/11/12193133/German-Shepherd-Puppy-Fetch-500x500.jpg",
  "https://images.glaciermedia.ca/polopoly_fs/1.24101709.1584641750!/fileImage/httpImage/image.jpg_gen/derivatives/landscape_804/rcmp-dog-puppies-police-dog-services-training-centre-2020.jpg",
  "https://static.boredpanda.com/blog/wp-content/uploads/2018/10/cutest-puppy-dog-pictures-coverimage.jpg",
  "https://media.mnn.com/assets/images/2018/10/DqM7GXEU8AATAKL.jpg.838x0_q80.jpg",
  "https://www.cp24.com/polopoly_fs/1.4966708.1591148397!/httpImage/image.jpg_gen/derivatives/landscape_960/image.jpg",
  "https://www.urbaneer.com/images/resized/media/uploads/blog/puppy_yoga_600x450.jpg",
  "https://www.thelabradorsite.com/wp-content/uploads/2019/04/When-Do-Puppies-Stop-Biting-And-How-To-Cope-With-A-Teething-Lab-Puppy-LS-long.jpg",
  "https://vetmed.illinois.edu/wp-content/uploads/2019/06/pc-vitt-murmur.jpg",
  "https://scx1.b-cdn.net/csz/news/800/2020/scientistsun.jpg",
  "https://cdn.cnn.com/cnnnext/dam/assets/200217140048-02-puppy-pigeon-friendship-rescue-large-169.jpg",
  "https://www.browntrout.ca/media/catalog/product/cache/1/image/9df78eab33525d08d6e5fb8d27136e95/9/7/9781975412654.MAIN.jpg",
  "https://www.cbc.ca/kidscbc2/content/the_feed/_848/puppies-HEADER-compressor.jpg",
  "https://4evaidrg3wj26w4ml1bpc9el-wpengine.netdna-ssl.com/wp-content/uploads/2019/05/16645970_web1_190503-ABB-Stolen-puppies_2.jpg",
  "https://44fov5194kmkld3jc3wsk29c-wpengine.netdna-ssl.com/wp-content/uploads/2020/05/21564761_web1_200515-WLT-BCSPCA-Orphaned-Puppies_1.jpg",
  "https://www.saobserver.net/wp-content/uploads/2020/03/20835539_web1_200311-PWN-Huskies-HuskyPuppies_1.jpg",
  "https://cdn2-www.dogtime.com/assets/uploads/gallery/30-impossibly-cute-puppies/impossibly-cute-puppy-2.jpg",
  "https://www.journalpioneer.com/media/photologue/photos/cache/PE-17122019-Husky_puppy_large.jpg",
  "https://petcentral.chewy.com/wp-content/uploads/Facts-About-Puppies-300x168.jpg",
  "https://i.pinimg.com/originals/e5/d7/e0/e5d7e0dd5dc3b5925b4f59438e1a4dca.jpg",
  "https://www.akcpetinsurance.com/res/akc/blog/2017/the-truth-about-getting-two-puppies-at-once/header_two_puppies.jpg",
  "https://shawglobalnews.files.wordpress.com/2019/12/puppies1g.png",
  "https://www.ctvnews.ca/polopoly_fs/1.4734947.1576677009!/httpImage/image.jpg_gen/derivatives/landscape_960/image.jpg",
  "https://infotel.ca/news/medialibrary/image/hd-mediaitemid73312-9802.png",
  "https://s3.amazonaws.com/cdn-origin-etr.akc.org/wp-content/uploads/2016/07/15155435/Pug-puppies.464864288-1.jpg",
  "https://media.socastsrm.com/wordpress/wp-content/blogs.dir/2311/files/2019/12/01bc1e13-39ac-4b48-b921-31e7efb3aa90.png",
  "https://dynamicmedia.zuza.com/zz/m/original_/b/c/bc6132a1-e834-46c3-82df-57561dae00a8/B881112906Z.1_20200420180303_000_GSOUEIIR.3-0_Super_Portrait.jpg",
  "https://usercontent1.hubstatic.com/14298682_f520.jpg",
  "https://i.ytimg.com/vi/8sUOvDzmrks/hqdefault.jpg",
  "https://www.thenorthernview.com/wp-content/uploads/2020/01/20048015_web1_200107-PWN-puppies.jpg",
]
`;
